### ISSUE TITLE
forbid access to file system to mini crates

### ISSUE THREAD

**Issue by LucianoBestia** [rust-secure-code/wg#37]
The tree of dependencies can be pretty deep.
It is nice to know that some crates have no unsafe code with `#![forbid(unsafe_code)]`. 
But should be nice also to know that they don't have access to files.
Maybe something like `#![forbid(file_access)]`.
If the crate needs read or write to files (like logging) than the parent crate should provide that.
So the parent crate has control what folders and files to allow and has the possibility to monitor what the mini crate is doing with files.
I know it is impossible to control file-access if there is FFI, assembler code or unsafe. 
But if the crate is already  #![forbid(unsafe_code)], maybe is possible to assure also `#![forbid(file_access)]` ?
That would make the Rust ecosystem with a lot of mini crate much more trustworthy.

**Comment by burdges** [comment]
Isn't network access as bad or worse?  I'd rephrase this:  Can we `#![forbid(std)]` so that builds fail if std gets linked?  

I'm dubious that security represents the best argument there, but std being quietly included create complexity throughout the ecosystem, so you'll receive enormous support if you find a technique that serves both down and up stream crate authors, as well as security interests.

**Comment by LucianoBestia** [comment]
ok, also network access could be harmful:  #![forbid(network_access)] 
I know that big crates need all that stuff. So we must be super careful who to trust.
But mini crates mostly don't need that. And could be trusted easily if the compiler can asserts they cannot be harmful (forbid unsafe, file read, file write, network access, build.rs, FFI, asm! ).

Yes, the choice to depend quietly on the standard library by default is strange for a language that is made for embedded and system programming. Sometimes they don't depend on any library, but must write #![no_std].
The other group of developers for CLI, servers, clients they usually depend on tens of libraries, but don't need to add the standard library there. So they save one line ;-)
If the standard library could be split in 3 pieces:
- one without file access and network
- one only for file access
- one only for network
That would be easier to tell what a crate can and cannot do. But maybe this is impossible. I really don't know.

**Comment by burdges** [comment]
You'd forbid specific dependencies from invoking specific other dependencies, like std, or their features.

**Comment by tarcieri** [comment]
Here was my take on this...

https://internals.rust-lang.org/t/crate-capability-lists/8933/2?u=bascule

## Unsafe Features

Synopsis: Extend the existing idea of cargo features with a special notion of "unsafe features" which can be used to whitelist usage of `unsafe` in dependencies (and their transitive dependencies).

### Goals

- Opt-in feature which has no effect on code that doesn't make use of the feature explicitly
- Cover all usages of `unsafe`, including `std`
- Provide a path for retrofitting `std` with unsafe features, which could eventually be used to eliminate ambient authority from Rust and thereby provide a foundation for an object capability model

### Non-Goals

- Full OCap semantics out of the box
- Breaking changes of any kind

### Cargo.toml changes

Let's start with something which is perhaps worthy of a pre-pre-RFC in and of itself, an `allow-unsafe` option:

```
[dependencies]
foobar = { version = "0.2", allow-unsafe = false }
```

This would *transitively* disallow use of `unsafe` by `foobar` and all of `foobar`'s transitive dependencies.

Ideally though, we could whitelist specific usages of unsafe in the `foobar` crate. Enter "unsafe features".

In the `Cargo.toml` for the `foobar` crate, we could imagine something like this:

```
[features.unsafe]
fire_the_missiles=[]
```

And in the code, something like this:

```
#[cfg(unsafe_feature = "fire_the_missiles")]
fn fire_the_missiles(...) {
   unsafe {
      // Use your imagination. How about some pointer arithmetic based on attacker-controlled data?
      [...]
   }
}
```

Now let's imagine the `baz` crate wants to consume the `foobar` crate and use this feature. It will need to opt into using this unsafe behavior (I am imaging that unsafe features **cannot** be default features, and must be opted into explicitly).

In the `Cargo.toml` for the `baz` crate, we do the following:

```
[dependencies]
foobar = { version = "0.2", unsafe-features = ["fire_the_missiles"] }
```

Now the `baz` crate is able to make use of the `fire_the_missiles` function. We can imagine that all of the rest of the cargo feature behavior continues to work, for example `foobar` could be an optional dependency, pulled in via a regular (non-unsafe) cargo feature.

But for simplicity's sake (and to illustrate an example), let's suppose that `baz` always includes the `foobar` crate and makes use of the `foobar/fire_the_missiles` unsafe feature. Now what happens when the `quux` crate tries to include `baz`?

In the `Cargo.toml` for the `quux` crate, imagine we did this:

```
[dependencies]
baz = "0.1"
```

This is where things get a bit interesting. `baz` is making use of the `foobar/fire_the_missiles` feature, and `quux` has not explicitly authorized it. This is an error:

```
error: crate `quux` makes use of `unsafe-feature` not whitelisted in Cargo.toml: foobar/fire_the_missiles`
```

...or thereabouts.

To correct this, we need to explicitly whitelist this relationship in the `quux` crate's `Cargo.toml`:

```
[dependencies]
baz = { version = "0.1", unsafe-features = ["foobar/fire_the_missiles"] }
```

Explicitly whitelisting these features would be required at *any* level.

Imagine we're in a completely different project which is including the `quux` crate. It would be required to do the following in its `Cargo.toml`

```
[dependencies]
quux = { version = "0.0.1", unsafe-features = ["baz/foobar/fire_the_missiles"] }
```

### Tough questions

I stated one of the goals is "opt-in feature which has no effect on code that doesn't make use of the feature explicitly". In my spitball description, I'm suggesting whitelisting of `unsafe` usages "kicks in" when `allow-unsafe = false` or `unsafe-features` is added to a crate's attributes in the `[dependencies]` section. But what about:

- Q1: Usages of `unsafe` which aren't tagged with `#[cfg(unsafe_feature)]`?
- Q2: Usages of `unsafe` in `std`

Well, short answers:

- A1: When `unsafe-features` is used, **all** usages of unsafe **MUST** be gated on a `#[cfg(unsafe_feature)]`. If that were the case, it would probably make sense to make these usages compile errors.
- A2: It should probably apply to `std`, possibly with an opt-out mechanism for truly side-effect and ambient authority-free things like mutating the interiors of strings.

This means to get the unsafe parts of `std` "back" in these crates and make them accessible, `std` itself would need to be retrofitted with `unsafe-feature`s.

A quick summary of the philosophy here:

- Some crates export `unsafe-feature`s
- Other crates consume them, explicitly whitelisting the ones they use
- At each level of the dependency hierarchy, crates must opt into these features, or these uses of unsafe will be a compile error

**Comment by LucianoBestia** [comment]
That sounds really good to me. 
I would like to emphasize that the word "unsafe" in Rust has this exclusive meaning for memory safety.  That is not how the word "unsafe" is understood by normal people.
We should find another word to be unambiguous like: "nothing is really stopping this mini crate to do harm"
I would add "file access" and "network access",  as the next in line of potentially harmful features.

**Comment by tarcieri** [comment]
> I would like to emphasize that the word "unsafe" in Rust has this exclusive meaning for memory safety. That is not how the word "unsafe" is understood by normal people.

I think you're missing the point: this uses `unsafe`, in the struct Rust memory safety sense, as invoked using the `unsafe` keyword, as the modeling dimension for such features.

Without `unsafe`, aside from a handful of other escape hatches like `#[no_mangle]`, a process can't make system calls and therefore can't do anything that involves accessing files, the network, the process environment, etc. All of these capabilities can ultimately be mediated through the use of the `unsafe` keyword as used by `std`.

**Comment by burdges** [comment]
You cannot allocate memory or build data structures without unsafe either though.  In this issue, I'd assume the goal was to permit `unsafe` where crates only use it for memory management, but forbiding io.  It's true std being dragged into builds has really plagued the ecosystem, so if you can forbid it nicely enough then it'll help various things.

**Comment by tarcieri** [comment]
The idea of "unsafe features" is to categorically tag usages of `unsafe` under a namespace of "unsafe features", so that you can control access to any subsystem, rather than building something specialized to a particular function.

**Comment by LucianoBestia** [comment]
yes, deep down file/network access is also unsafe.
Probably I wanted to express there is different levels of unsafe:
- There is the "pure unsafe" that can do literally anything. No-one can stop it.
- And there is "file access" that can do only file access. It is unsafe, but not so broad and unpredictable.
I think I am starting to categorize unsafe into specific categories that are specialized for something and cannot do other unexpected things. This is what a safe wrapper should do. If this wrapper is in a separate library. It can be used or forbidden by the "unsafe features". Great, it works.
I read that this was a long discussion and just wanted to check if there is something new and interesting.
I think that the article of Lin Clark
https://bytecodealliance.org/articles/announcing-the-bytecode-alliance
made me think. They want to solve the problem of file/net access only in WebAssembly. 
What about other Rust projects? Could it be done in compile time?

**Comment by LucianoBestia** [comment]
I looked at your text "Unsafe Features"again with other eyes now.
Instead of "launch missile" I see "read a file in a dangerous folder".
And all seems logic. You can forbid or allow it. Consciously opt-in.
I like it.

**Comment by tarcieri** [comment]
Not quite. Perhaps I need to do a better job explaining it.

The idea is if you can forbid all *transitive* uses of unsafe from a Rust program (i.e. when one crate includes another crate in its Cargo.toml and wants to restrict the authority of its dependencies), with a scant handful of exceptions (like `#[no_mangle]`) you eliminate all ambient authority: it can't access the disk, the network, the process environment, or even the heap. The code is effectively "safe".

If you can get the compiler to reject all transitive unsafe usages for a particular crate, you can then add them back by whitelisting them. Cargo features already provide a modeling dimension by which which we can whitelist code.

Using the proposed `#[cfg(unsafe_feature = "fire_the_missiles")]` attribute, a parent crate including a sub-crate as a dependency can whitelist transitive usages of `unsafe` in this manner.

**Comment by LucianoBestia** [comment]
I would understand it sooner if instead of fire_missile you have described how to limit a file access.
Not generic for every file, just for one single dangerous folder. That folder must be known and fixed in the code. So I can allow it on forbid it to read there with "unsafe features".
Let's say the crypto wallet, that is a modern topic now.
But now I understand also the fire missile example.

**Comment by tarcieri** [comment]
That's a bit different idea. To pull that off you'd need some sort of OCap notion for paths.

The "unsafe features" proposal would be more like inside of `std::fs`, you'd label types which perform unsafe filesystem accesses:

```rust
/// This is `std::fs::File`
#[cfg(unsafe_feature = "fs")]
pub struct File {
   ...
}
```

Then in the `Cargo.toml` of a crate which wants to bless a dependency to access the filesystem, something like:

```toml
[dependencies]
foobar = { version = "0.2", unsafe-features = ["std/fs"] }
```

**Comment by LucianoBestia** [comment]
I build a little demonstrative example for "unsafe features". 
@tarcieri, Can you give it a look ?
<https://github.com/LucianoBestia/unsafe_features_example>

**Comment by tarcieri** [comment]
@LucianoBestia sorry, haven't had time yet but I'll take a look soon

**Comment by LucianoBestia** [comment]
One idea is to make mandatory one parameter for unsafe block. That parameter is the feature.
```rust
unsafe("unsafe_feature_write_file"){
...
}  
```
The same feature could be used broadly in the same module, but not outside.

**Comment by LucianoBestia** [comment]
But one of the problems I hear is macros. They can have unsafe and is impossible to expand them in 100% working code. I looked at Geiger.
